### PR TITLE
Use SupportedOSPlatforms property to generate assembly attributes

### PR DESCRIPTION
**Pull request by jeffhandley**
In preparation for taking in the [Platform Compatibility Analyzer](https://github.com/dotnet/roslyn-analyzers/pull/3917), we need to address build errors that will be introduced by the analyzer due to our `[SupportedOSPlatform]` and `[UnsupportedOSPlatform]` annotations.

This **Draft PR** uses an MSBuild-based approach similar to what was put in place for `<UnsupportedOSPlatforms>` such that assembly-level attributes can be generated based on the platforms listed in that property; a `<SupportedOSPlatforms>` property is introduced that works the same way. This approach will prevent supported and unsupported combinations in assembly attributes.

### CODE REVIEW

**Comment by jeffhandley** [comment]
Here are the current build errors that occur when using a private build from https://github.com/dotnet/roslyn-analyzers/pull/3917.

```
Build FAILED.

System\Net\Sockets\SocketAsyncEventArgs.Windows.cs(1222,60): error CA1416: 'ThreadPoolBoundHandle.GetNativeOverlappedState(NativeOverlapped*)' is supported on 'windows' [C:\Users\jeffhand\git\dotnet\runtime-master-supportedosplatforms\src\libraries\System.Net.Sockets\src\System.Net.Sockets.csproj]
C:\Users\jeffhand\git\dotnet\runtime-master-supportedosplatforms\src\libraries\Common\src\Interop\Windows\WinSock\SafeNativeOverlapped.cs(66,33): error CA1416: 'ThreadPoolBoundHandle.FreeNativeOverlapped(NativeOverlapped*)' is supported on 'windows' [C:\Users\jeffhand\git\dotnet\runtime-master-supportedosplatforms\src\libraries\System.Net.Sockets\src\System.Net.Sockets.csproj]
C:\Users\jeffhand\git\dotnet\runtime-master-supportedosplatforms\src\libraries\Common\src\System\Net\ContextAwareResult.Windows.cs(16,32): error CA1416: 'WindowsIdentity.GetCurrent()' is supported on 'windows' [C:\Users\jeffhand\git\dotnet\runtime-master-supportedosplatforms\src\libraries\System.Net.Sockets\src\System.Net.Sockets.csproj]
C:\Users\jeffhand\git\dotnet\runtime-master-supportedosplatforms\src\libraries\Common\src\System\Net\ContextAwareResult.Windows.cs(23,17): error CA1416: 'WindowsIdentity.Dispose()' is supported on 'windows' [C:\Users\jeffhand\git\dotnet\runtime-master-supportedosplatforms\src\libraries\System.Net.Sockets\src\System.Net.Sockets.csproj]
System\Net\Sockets\SafeSocketHandle.Windows.cs(108,17): error CA1416: 'ThreadPoolBoundHandle.Dispose()' is supported on 'windows' [C:\Users\jeffhand\git\dotnet\runtime-master-supportedosplatforms\src\libraries\System.Net.Sockets\src\System.Net.Sockets.csproj]
System\Net\Sockets\SafeSocketHandle.Windows.cs(79,74): error CA1416: 'ThreadPoolBoundHandle.Handle' is supported on 'windows' [C:\Users\jeffhand\git\dotnet\runtime-master-supportedosplatforms\src\libraries\System.Net.Sockets\src\System.Net.Sockets.csproj]
System\Net\Sockets\SafeSocketHandle.Windows.cs(54,39): error CA1416: 'ThreadPoolBoundHandle.BindHandle(SafeHandle)' is supported on 'windows' [C:\Users\jeffhand\git\dotnet\runtime-master-supportedosplatforms\src\libraries\System.Net.Sockets\src\System.Net.Sockets.csproj]
System\Net\Sockets\SocketAsyncEventArgs.Windows.cs(84,20): error CA1416: 'ThreadPoolBoundHandle.AllocateNativeOverlapped(PreAllocatedOverlapped)' is supported on 'windows' [C:\Users\jeffhand\git\dotnet\runtime-master-supportedosplatforms\src\libraries\System.Net.Sockets\src\System.Net.Sockets.csproj]
System\Net\Sockets\SocketAsyncEventArgs.Windows.cs(96,13): error CA1416: 'ThreadPoolBoundHandle.FreeNativeOverlapped(NativeOverlapped*)' is supported on 'windows' [C:\Users\jeffhand\git\dotnet\runtime-master-supportedosplatforms\src\libraries\System.Net.Sockets\src\System.Net.Sockets.csproj]
System\Net\Sockets\SocketAsyncEventArgs.Windows.cs(60,43): error CA1416: 'PreAllocatedOverlapped' is supported on 'windows' [C:\Users\jeffhand\git\dotnet\runtime-master-supportedosplatforms\src\libraries\System.Net.Sockets\src\System.Net.Sockets.csproj]
System\Net\Sockets\SocketAsyncEventArgs.Windows.cs(907,17): error CA1416: 'PreAllocatedOverlapped.Dispose()' is supported on 'windows' [C:\Users\jeffhand\git\dotnet\runtime-master-supportedosplatforms\src\libraries\System.Net.Sockets\src\System.Net.Sockets.csproj]
System\Net\Sockets\BaseOverlappedAsyncResult.Windows.cs(53,48): error CA1416: 'ThreadPoolBoundHandle.AllocateNativeOverlapped(IOCompletionCallback, object?, object?)' is supported on 'windows' [C:\Users\jeffhand\git\dotnet\runtime-master-supportedosplatforms\src\libraries\System.Net.Sockets\src\System.Net.Sockets.csproj]
System\Net\Sockets\BaseOverlappedAsyncResult.Windows.cs(62,80): error CA1416: 'ThreadPoolBoundHandle.GetNativeOverlappedState(NativeOverlapped*)' is supported on 'windows' [C:\Users\jeffhand\git\dotnet\runtime-master-supportedosplatforms\src\libraries\System.Net.Sockets\src\System.Net.Sockets.csproj]
    0 Warning(s)
    13 Error(s)
```

File: src/libraries/Directory.Build.targets [hunk 1]
```diff
@@ -243,7 +243,7 @@
 
   <!-- Adds System.Runtime.Versioning*Platform* annotation attributes to < .NET 5 builds -->
   <Choose>
-    <When Condition="('$(IncludePlatformAttributes)' == 'true' or '$(IsWindowsSpecific)' == 'true') and ('$(TargetFrameworkIdentifier)' != '.NETCoreApp' or $([MSBuild]::VersionLessThan($(TargetFrameworkVersion), '5.0')))">
+    <When Condition="('$(IncludePlatformAttributes)' == 'true' or '$(SupportedOSPlatforms)' != '' or '$(UnsupportedOSPlatforms)' != '') and ('$(TargetFrameworkIdentifier)' != '.NETCoreApp' or $([MSBuild]::VersionLessThan($(TargetFrameworkVersion), '5.0')))">
```
Comment (adamsitnik) [inline src/libraries/Directory.Build.targets:246]: very good idea to not always require `IncludePlatformAttributes` when `SupportedOSPlatforms` or `UnsupportedOSPlatforms` are used 👍 

You could simplify the assemblies annotated in https://github.com/dotnet/runtime/pull/40612 and replace

```xml
    <IncludePlatformAttributes>true</IncludePlatformAttributes>
    <UnsupportedOSPlatforms>browser</UnsupportedOSPlatforms>
```

with just

```xml
    <UnsupportedOSPlatforms>browser</UnsupportedOSPlatforms>
```

**Review by adamsitnik — APPROVED** [review]
Very good idea! LGTM 👍

File: eng/versioning.targets [hunk 2]
```diff
@@ -22,20 +22,18 @@
...
     <_unsupportedOSPlatforms Include="$(UnsupportedOSPlatforms)" />
   </ItemGroup>
 
-  <Target Name="AddUnsupportedOSPlatformAttribute" BeforeTargets="GenerateAssemblyInfo" AfterTargets="PrepareForBuild">
+  <Target Name="AddAssemblyOSPlatformAttributes" BeforeTargets="GenerateAssemblyInfo" AfterTargets="PrepareForBuild">
     <ItemGroup>
-      <AssemblyAttribute Include="System.Runtime.Versioning.UnsupportedOSPlatform" Condition="'@(_unsupportedOSPlatforms)' != ''">
+      <AssemblyAttribute Include="System.Runtime.Versioning.SupportedOSPlatform" Condition="'@(_supportedOSPlatforms)' != ''">
+        <_Parameter1>%(_supportedOSPlatforms.Identity)</_Parameter1>
+      </AssemblyAttribute>
```
Comment (terrajobst) [inline eng/versioning.targets:35]: I don't think you need the condition here because you use `%()` to for each over the item group.

File: eng/versioning.targets [hunk 3]
```diff
@@ -22,20 +22,18 @@
...
+      <AssemblyAttribute Include="System.Runtime.Versioning.UnsupportedOSPlatform" Condition="'@(_supportedOSPlatforms)' == '' and '@(_unsupportedOSPlatforms)' != ''">
         <_Parameter1>%(_unsupportedOSPlatforms.Identity)</_Parameter1>
       </AssemblyAttribute>
```
Comment (terrajobst) [inline eng/versioning.targets:38]: I think you can remove the second half of the condition.

Comment (terrajobst) [inline eng/versioning.targets:38]: Should we produce a build warning when both `$(SupportedOSPlatforms)` and `$(UnsupportedOSPlatforms)` are set? Those are mutual exclusive choices, right?

**Review by terrajobst — APPROVED** [review]

**Comment by ViktorHofer** [comment]
@jeffhandley curious, is this runtime specific or should customers be able to set this in their application as well? If the latter this would probably need to go into dotnet/sdk.

**Comment by jeffhandley** [comment]
> @jeffhandley curious, is this runtime specific or should customers be able to set this in their application as well? If the latter this would probably need to go into dotnet/sdk.

This is specific to the runtime repo, making it easier for us to apply assembly-wide attributes throughout our repo in various configs, but ensuring we don't produce conflicting `[SupportedOSPlatform]` and `[UnsupportedOSPlatform]` attributes.  We could consider productizing this in the SDK in the future though.

**Comment by ViktorHofer** [comment]
Let me rephrase. Will the platform compat analyzer which respects these attributes analyzer customer applications as well or only runtime libraries?

**Comment by jeffhandley** [comment]
Ah, thanks for rephrasing.

The analyzer will be part of the SDK through the NetAnalyzers package.  https://github.com/dotnet/roslyn-analyzers/pull/3917 will add the analyzer into `roslyn-analyzers`, and then there will be a step to update the SDK with the latest `roslyn-analyzers` build.

This PR sets the runtime repo up to be able to mark entire assemblies with `[SupportedOSPlatform]` attributes to make the assemblies platform-specific where appropriate, so that we can be ready for the intake of that analyzer for our repo.

**Comment by eerhardt** [comment]
I had a similar initial reaction as @ViktorHofer: Should this be an official product feature? If this is generally applicable to our customers, then we should probably add the capability to dotnet/sdk instead of our build infrastructure. However, I can understand if it is too late in 5.0 to get that in now - but maybe something in 6.0?

File: src/libraries/Microsoft.Win32.Registry.AccessControl/Directory.Build.props [hunk 4]
```diff
@@ -2,6 +2,6 @@
   <Import Project="..\Directory.Build.props" />
   <PropertyGroup>
     <StrongNameKeyId>Microsoft</StrongNameKeyId>
-    <IsWindowsSpecific>true</IsWindowsSpecific>
+    <SupportedOSPlatforms>windows</SupportedOSPlatforms>
```
Comment (eerhardt) [inline src/libraries/Microsoft.Win32.Registry.AccessControl/Directory.Build.props:5]: I thought `SupportedOSPlatforms` was an Item and not a property?

Comment (jeffhandley) [reply src/libraries/Microsoft.Win32.Registry.AccessControl/Directory.Build.props:5]: There are 2 with similar naming:

- `<SupportedPlatform>`, which is an item, and it tells the analyzer which platforms we care to see warnings about, and it's part of the SDK product
- `<SupportedOSPlatform>`, which is this property, and it results in assembly-level `[SupportedOSPlatform]` attributes, and it's part of the runtime repo only

These names are so similar they're surely easy to get mixed up.  Perhaps a better name for the runtime repo properties that produce the assembly-level attributes would be:

- `<AssemblySupportedOSPlatformAttributes>`
- `<AssemblyUnsupportedOSPlatformAttributes>`

**Comment by bartonjs** [comment]
> I had a similar initial reaction as @ViktorHofer: Should this be an official product feature?

My understanding is that anyone building with a platform-specific TFM (e.g. net5.0-win) will have an assembly-level attribute automatically applied (whether that work is done or not, I don't know, but it was discussed).  In our repository we build with the platform agnostic TFM (net5.0) for our OS-split assemblies.  So I think we're just back-plumbing that SDK feature into our build variant.

Another possibility would be to try building our OS-split libraries with their OS-variant TFM, but I have no idea what impact that would have on our build composition.  It's maybe the right answer for 6, but probably risky for 5 (unless it Just Works).

File: eng/versioning.targets (hunk 3 above)
Comment (jeffhandley) [reply eng/versioning.targets:38]: Good idea

Comment (safern) [reply eng/versioning.targets:38]: Yeah we don't need `_supportedOSPlatforms condition.

File: eng/versioning.targets (hunk 2 above)
Comment (safern) [reply eng/versioning.targets:35]: You do need the condition because if the item is empty it will add an attribute without `Parameter1`

File: eng/versioning.targets (hunk 3 above)
Comment (jeffhandley) [reply eng/versioning.targets:38]: Ah, good catch. Thanks.

File: eng/versioning.targets (hunk 2 above)
Comment (jeffhandley) [reply eng/versioning.targets:35]: Ah, good catch. Thanks.

File: eng/versioning.targets (hunk 3 above)
Comment (jeffhandley) [reply eng/versioning.targets:38]: I was checking `_supportedOSPlatforms` here to prevent any `[UnsupportedOSPlatform]` attributes from occurring when there are already `[SupportedOSPlatform]` attributes, as we must only emit one or the other (with _Supported_ being more restrictive and therefore taking precedence).

I'll take the suggestion of emitting a build warning when this conflict occurs, but I recommend we keep the check in place as having both attributes in place can lead to lots of other noise in the compilation, including even crashes in the analyzer runner at the moment.

Comment (safern) [reply eng/versioning.targets:38]: Makes sense. However we should be able to catch these kind of issues on PR review? And the warning makes sense.

File: src/libraries/Microsoft.Extensions.Hosting/src/Host.cs [hunk 5]
```diff
@@ -111,8 +111,11 @@ public static IHostBuilder CreateDefaultBuilder(string[] args)
 
                 if (isWindows)
                 {
+#pragma warning disable CA1416 // Platform compat analyzer
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Hosting/src/Host.cs:114]: Why is this necessary? `isWindows` is defined above as:

```
bool isWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
```

Can the analyzer not recognize this pattern?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPListener.cs [hunk 6]
```diff
@@ -274,7 +275,7 @@ private void CreateNewSocketIfNeeded()
                 _serverSocket.ExclusiveAddressUse = true;
             }
 
-            if (_allowNatTraversal != null)
+            if (_allowNatTraversal != null && OperatingSystem.IsWindows())
```
Comment (eerhardt) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPListener.cs:278]: Are we concerned about behavior change here at all? Previously someone would get a `PNSE` during socket creation, if they called `TCPListener.AllowNatTraversal(bool allowed)`. Now it looks like it will no longer PNSE.

File: eng/versioning.targets [hunk 7]
```diff
@@ -24,19 +24,23 @@
...
+      <!-- When there are Supported OSPlatforms, transform each of those into an attribute -->
       <AssemblyAttribute Include="System.Runtime.Versioning.SupportedOSPlatform" Condition="'@(_supportedOSPlatforms)' != ''">
         <_Parameter1>%(_supportedOSPlatforms.Identity)</_Parameter1>
       </AssemblyAttribute>
+      <!-- When there are not any Supported OSPlatforms, but there are Unsupported OSPlatforms, transform each of those into an attribute -->
       <AssemblyAttribute Include="System.Runtime.Versioning.UnsupportedOSPlatform" Condition="'@(_supportedOSPlatforms)' == '' and '@(_unsupportedOSPlatforms)' != ''">
         <_Parameter1>%(_unsupportedOSPlatforms.Identity)</_Parameter1>
       </AssemblyAttribute>
     </ItemGroup>
+    <Warning Condition="'@(_supportedOSPlatforms)' != '' and '@(_unsupportedOSPlatforms)' != ''"
```
Comment (safern) [inline eng/versioning.targets:42]: I would just Error. We set warnAsError for the global build, so it is the same, we usually try to error for this kind of stuff as warnings could slip sometimes.

**Comment by jeffhandley** [comment]
We're still iterating on this; I'm going to close the draft PR and then we will open another PR when ready for another review.  The feedback here was helpful to steer the direction; thanks!

File: src/libraries/Microsoft.Extensions.Hosting/src/Host.cs (hunk 5 above)
Comment (buyaa-n) [reply src/libraries/Microsoft.Extensions.Hosting/src/Host.cs:114]: Analyzer can track result save in local, it shouldn't warn

Comment (jeffhandley) [reply src/libraries/Microsoft.Extensions.Hosting/src/Host.cs:114]: It was warning on this.  Remind me; can the analyzer also recognize the old `RuntimeInformation.IsOSPlatform()` calls like this one? This branch is still intact despite the PR being close.  Could you pull the branch down and check this scenario please, @buyaa-n?

Comment (buyaa-n) [reply src/libraries/Microsoft.Extensions.Hosting/src/Host.cs:114]: > can the analyzer also recognize the old RuntimeInformation.IsOSPlatform()

Yes it can

> It was warning on this. 

That is strange, i added exact same test and it is not warning

> This branch is still intact despite the PR being close. Could you pull the branch down and check this scenario please, @buyaa-n?

Sure

Comment (buyaa-n) [reply src/libraries/Microsoft.Extensions.Hosting/src/Host.cs:114]: Filed an issue for the bug https://github.com/dotnet/runtime/issues/41535
